kotlin/workflow-ui-android/src/main/java/com/squareup/workflow/ui/WorkflowViewModel.kt

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2019 Square Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.workflow.ui
+
+import android.arch.lifecycle.ViewModel
+import android.arch.lifecycle.ViewModelProvider
+import com.squareup.workflow.Snapshot
+import com.squareup.workflow.Workflow
+import com.squareup.workflow.WorkflowHost
+import io.reactivex.android.schedulers.AndroidSchedulers
+import io.reactivex.disposables.Disposable
+import kotlinx.coroutines.experimental.Dispatchers
+import kotlinx.coroutines.experimental.rx2.asObservable
+
+/**
+ * The guts of [WorkflowActivityRunner]. We could have made that class itself a
+ * [ViewModel], but that would allow accidental calls to [onCleared], which
+ * would be nasty.
+ */
+internal class WorkflowViewModel<OutputT : Any, RenderingT : Any>(
+  val viewRegistry: ViewRegistry,
+  host: WorkflowHost<OutputT, RenderingT>
+) : ViewModel() {
+
+  internal class Factory<InputT : Any, OutputT : Any, RenderingT : Any>(
+    private val viewRegistry: ViewRegistry,
+    private val workflow: Workflow<InputT, OutputT, RenderingT>,
+    private val initialInput: InputT,
+    private val restored: PickledWorkflow?
+  ) : ViewModelProvider.Factory {
+    private val hostFactory = WorkflowHost.Factory(Dispatchers.Unconfined)
+
+    override fun <T : ViewModel> create(modelClass: Class<T>): T {
+      val host = hostFactory.run(workflow, initialInput, restored?.snapshot)
+      @Suppress("UNCHECKED_CAST")
+      return WorkflowViewModel(viewRegistry, host) as T
+    }
+  }
+
+  private lateinit var sub: Disposable
+
+  var lastSnapshot: Snapshot = Snapshot.EMPTY
+
+  val updates =
+    host.updates.asObservable(Dispatchers.Unconfined)
+        // Unclear to me why this is necessary.
+        .observeOn(AndroidSchedulers.mainThread())
```

Comment:
@zach-klippenstein particularly wondering why I crash w/view operations from the wrong thread w/o this. What's taking us off of main?

Reply:
Network ops? We're still using `Dispatchers.Unconfined` everywhere, so if you're waiting for an network call to finish on a background thread, it could move us over.

Try changing your dispatcher to `Dispatchers.Main.immediate`.

Reply:
Not available on this branch. I'll file a ticket.

Reply:
#252